File: 김신영/5주차/14889.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    int answer;
    int n;
    int[][] abilityArray;
    int arraySum = 0;
    int halfN;

    public int solution() throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer stringTokenizer = new StringTokenizer(br.readLine());

        n = Integer.parseInt(stringTokenizer.nextToken());
        halfN = n / 2;
        abilityArray = new int[n][n];

        for (int i = 0; i < n; i++) {
            stringTokenizer = new StringTokenizer(br.readLine());
            for (int j = 0; i < n; j++) {
                if (stringTokenizer.hasMoreTokens()) {
                    int temp = Integer.parseInt(stringTokenizer.nextToken());
                    abilityArray[i][j] = temp;
                    arraySum += temp;
                } else {
                    break;
                }
                //여기 왜 이래
            }
        }

        answer = arraySum;
        boolean[] visited = new boolean[n];
        visited[0] = true;
        combination(1, visited, 1);

        return answer;
    }

    private void combination(int depth, boolean[] visited, int index) {
        if (depth == halfN) {
            int start = 0;
            int link = 0;

            for (int i = 0; i < n; i++) {
                for (int j = i + 1; j < n; j++) {
                    int ij = abilityArray[i][j];
                    int ji = abilityArray[j][i];
                    if (!visited[i] && !visited[j]) {
                        link += ji + ij;
                    } else if (visited[i] && visited[j]) {
                        start += ji + ij;
                    }
                }
            }
            int diff = Math.abs(start - link);
            answer = Math.min(answer, diff);

            return;
        }

        for (int i = index; i < n; i++) {
            if (!visited[i]) {
                visited[i] = true;
                depth += 1;
                index = i;
                combination(depth, visited, index);
                visited[i] = false;
                depth -= 1;
            }
        }
    }

    public static void main(String[] args) throws Exception {
        Main startAndLink = new Main();
        System.out.println(startAndLink.solution());
    }
}

File: 전효정/5주차/14889.py
# 아무리 노력해도 시간 안에 아이디어를 떠올릴 수 없을 것 같아
# 남의 코드를 가져와 외웠습니다.
# https://wlstyql.tistory.com/170

import sys
N = int(sys.stdin.readline())
_map = [list(map(int, sys.stdin.readline().split())) for _ in range(N)]
visited = [False] * N
result = []
_min = 1e9

def diff():
    _start = 0
    _link = 0
    for i in range(N-1):
        for j in range(i+1, N):
            #True일 때 스타트 팀
            if visited[i] and visited[j]:
                _start += _map[i][j]
                _start += _map[j][i]
            #False일 때 링크 팀
            elif not visited[i] and not visited[j]:
                _link += _map[i][j]
                _link += _map[j][i]
    return abs(_start - _link)

def dfs(depth, idx, N):
    global _min
    if depth == N/2:  #절반만 선택
        diff_result = diff()
        _min = min(_min, diff_result)
        if _min == 0:
            print(_min)
            exit(0)
        return
    
    for i in range(idx, N):
        if not visited[i]:
            visited[i] = True
            dfs(depth + 1, i + 1, N)
            visited[i] = False

    return

dfs(0, 0, N)
print(_min)

            
        
        
